src/streamable.rs

```diff
@@ -0,0 +1,625 @@
+use crate::chia_error::{Error, Result};
+use std::convert::TryInto;
+use std::mem::size_of;
+
+pub struct InputStream<'a> {
+    buffer: &'a [u8],
+    pos: usize,
+}
+
+impl<'a> InputStream<'a> {
+    pub fn new(buffer: &'a [u8]) -> Self {
+        InputStream { buffer, pos: 0 }
+    }
+
+    pub fn pos(&self) -> usize {
+        self.pos
+    }
+
+    pub fn read(&mut self, len: usize) -> Result<&'a [u8]> {
+        if self.buffer.len() - self.pos < len {
+            Err(Error::EndOfBuffer)
+        } else {
+            let ret = &self.buffer[self.pos..self.pos + len];
+            self.pos += len;
+            Ok(ret)
+        }
+    }
+}
+
+pub trait Streamable {
+    fn stream(&self, out: &mut Vec<u8>) -> Result<()>;
+    fn parse(input: &mut InputStream) -> Result<Self>
+    where
+        Self: Sized;
+}
+
+macro_rules! streamable_primitive {
+    ($t:ty) => {
+        impl Streamable for $t {
+            fn stream(&self, out: &mut Vec<u8>) -> Result<()> {
+                Ok(out.extend_from_slice(&self.to_be_bytes()))
+            }
+            fn parse(input: &mut InputStream) -> Result<Self> {
+                let sz = size_of::<$t>();
+                Ok(<$t>::from_be_bytes(input.read(sz)?.try_into().unwrap()))
+            }
+        }
+    };
+}
+
+streamable_primitive!(u8);
+streamable_primitive!(i8);
+streamable_primitive!(u16);
+streamable_primitive!(i16);
+streamable_primitive!(u32);
+streamable_primitive!(i32);
+streamable_primitive!(u64);
+streamable_primitive!(i64);
+
+impl<T: Streamable> Streamable for Vec<T> {
+    fn stream(&self, out: &mut Vec<u8>) -> Result<()> {
+        if self.len() > u32::MAX as usize {
+            Err(Error::InputTooLarge)
+        } else {
+            (self.len() as u32).stream(out)?;
+            for e in self {
+                e.stream(out)?;
+            }
+            Ok(())
+        }
+    }
+
+    fn parse(input: &mut InputStream) -> Result<Self> {
+        let len = u32::parse(input)?;
+
+        // TODO: pre-allocate capacity, but we'd need safe-guards for overflow
+        // attacks
+        let mut ret = Vec::<T>::new();
+        for _ in 0..len {
+            ret.push(T::parse(input)?);
+        }
+        Ok(ret)
+    }
+}
+
+impl Streamable for String {
```

Comment:
Maybe a little more challenging, but you might be able to use `Cursor` here too. You'd have to convert `chia_error::Error` to `std::io::Error` (and downcasting might lose some diagnostics, although I'd guess at this point all you get is a programmer-readable error). Might be no tangible benefit.